src/Features/Core/Portable/MakeFieldReadonly/MakeFieldReadonlyDiagnosticAnalyzer.cs

```diff
@@ -151,57 +151,46 @@ void UpdateFieldStateOnWrite(IFieldSymbol field)
-                fieldReference.Parent is IIncrementOrDecrementOperation ||
-                fieldReference.Parent is IArgumentOperation argumentOperation && argumentOperation.Parameter.RefKind != RefKind.None ||
-                IsInLeftOfDeconstructionAssignment(fieldReference))
+            // Check if the underlying member is being written or a writable reference to the member is taken.
+            var valueUsageInfo = fieldReference.GetValueUsageInfo();
+            if ((valueUsageInfo & (ValueUsageInfo.Write | ValueUsageInfo.WritableRef)) == 0)
             {
-                // Writes to fields inside constructor are ignored, except for the below cases:
-                //  1. Instance reference of an instance field being written is not the instance being initialized by the constructor.
-                //  2. Field is being written inside a lambda or local function.
-
-                // Check if we are in the constructor of the containing type of the written field.
-                var isInConstructor = owningSymbol.IsConstructor();
-                var isInStaticConstructor = owningSymbol.IsStaticConstructor();
-                var field = fieldReference.Field;
-                if ((isInConstructor || isInStaticConstructor) &&
-                    field.ContainingType == owningSymbol.ContainingType)
-                {
-                    // For instance fields, ensure that the instance reference is being initialized by the constructor.
-                    var instanceFieldWrittenInCtor = isInConstructor &&
-                        fieldReference.Instance?.Kind == OperationKind.InstanceReference &&
-                        (!isFieldAssignemnt || fieldReference.Parent.Parent?.Kind != OperationKind.ObjectOrCollectionInitializer);
+                return false;
+            }
+
+            // Writes to fields inside constructor are ignored, except for the below cases:
+            //  1. Instance reference of an instance field being written is not the instance being initialized by the constructor.
+            //  2. Field is being written inside a lambda or local function.
+
+            // Check if we are in the constructor of the containing type of the written field.
+            var isInConstructor = owningSymbol.IsConstructor();
+            var isInStaticConstructor = owningSymbol.IsStaticConstructor();
+            var field = fieldReference.Field;
+            if ((isInConstructor || isInStaticConstructor) &&
+                field.ContainingType == owningSymbol.ContainingType)
+            {
+                // For instance fields, ensure that the instance reference is being initialized by the constructor.
+                var instanceFieldWrittenInCtor = isInConstructor &&
+                    fieldReference.Instance?.Kind == OperationKind.InstanceReference &&
+                    (!(fieldReference.Parent is IAssignmentOperation) || fieldReference.Parent.Parent?.Kind != OperationKind.ObjectOrCollectionInitializer);
```

Comment:
this expression has gotten too complex to understand.  Far too many `&&`s, `||`s, and `!`s.